Add minimal support for Microchip SAMA7G54-EK

TonyHan11 opened the pull request:
- Add minimal support for sama7g54-ek as following

|Interface | Controller | Driver/Component|
|------|------|------|
| GICv2 | on-chip | arch/arm |
| GPIO | on-chip | pinctrl |
| PMC | on-chip | clock_control |
| SCKC | on-chip | clock_control |
| Flexcom | on-chip | mfd |
| USART |	on-chip | serial |
| PIT64B | on-chip | timer |

Need: zephyrproject-rtos/hal_microchip#27

TonyHan11 committed:
soc: microchip: add new soc sama7g54

Product URL: https://www.microchip.com/en-us/product/SAMA7G54

Signed-off-by: Tony Han <tony.han@microchip.com>

TonyHan11 committed:
boards: microchip: add new board sama7g54-ek

Product URL: https://www.microchip.com/en-us/development-tool/ev21h18a

Signed-off-by: Tony Han <tony.han@microchip.com>

TonyHan11 committed:
soc: microchip: sam: add code for sama7g5 clocks

Add code for sama7g5 Generic Clock, Main Clock, Main System Bus Clock,
Peripheral Clock, Programmable Clock and PLL Clock.

Signed-off-by: Tony Han <tony.han@microchip.com>

TonyHan11 committed:
drivers: serial: sam: update serial to support sama7g5

The USART is compatiable with the USART feature in sama7g5's Flexcom.
Update serial/usart_sam to support sama7g5.

Signed-off-by: Tony Han <tony.han@microchip.com>

TonyHan11 committed:
drivers: pinctrl: sam: add pinctrl for sama7g5

Support pull up/down, open drain for sam7g5's PIO.

Signed-off-by: Tony Han <tony.han@microchip.com>

TonyHan11 committed:
drivers: timer: sam: add pit64b timer for sama7g5

Use pit64b0 as systick, make "kernel sleep/uptime" commands work.

Signed-off-by: Tony Han <tony.han@microchip.com>

TonyHan11 committed:
dts: microchip: add the DTSI file for sama7g5

Add the base DTSI file for Microchip sama7g5.

Signed-off-by: Tony Han <tony.han@microchip.com>

TonyHan11 committed:
boards: microchip: sam: add the device-tree file for sama7g54_ek board

Introduce sama7g54_ek.dts for sama7g54_ek board.

Signed-off-by: Tony Han <tony.han@microchip.com>

TonyHan11 committed:
drivers: clock_control: microchip: add drivers for sama7g5 PMC

Initialize the configurations for PMC driver.
Add implement of the API for PMC clocks.

Signed-off-by: Tony Han <tony.han@microchip.com>

TonyHan11 committed:
drivers: clock_control: microchip: add support for sama7g5 SCKC

Add driver for sama7g5 Slow Clock Controller (SCKC).

Signed-off-by: Tony Han <tony.han@microchip.com>

TonyHan11 committed:
drivers: mfd: add FLEXCOM drivers for Microchip SAM devices

The FLEXCOM offers several serial communication protocols that are
managed by the three sub-modules USART, SPI, and TWI (I2C).

Signed-off-by: Tony Han <tony.han@microchip.com>

TonyHan11 commented on drivers/pinctrl/pinctrl_sam_pio4.c:
```diff
@@ -0,0 +1,64 @@
+	}
+	if (pin.bias_pull_up) {
+		conf |= PIO_CFGR_PUEN(PIO_CFGR_PUEN_ENABLED_Val);
+		conf &= PIO_CFGR_PDEN_Msk;
```
Updated with the fix.
`conf &= PIO_CFGR_PDEN_Msk;` changed to `conf &= ~PIO_CFGR_PDEN_Msk;`
`conf &= PIO_CFGR_PUEN_Msk;` changed to `conf &= ~PIO_CFGR_PUEN_Msk;`.

nordicjm commented on soc/microchip/sam/Kconfig.defconfig:
```diff
@@ -0,0 +1,17 @@
+	default y
+
+config MFD
+	default y
+	depends on DT_HAS_MICROCHIP_SAMA5D2_FLEXCOM_ENABLED
```
the underlying driver should select MFD since it requires it, not do this

TonyHan11 replied:
OK, the config here will be removed and add `select MFD` in `config USART_SAM`. Thanks.

nordicjm commented on boards/microchip/sam/sama7g54_ek/Kconfig.defconfig:
```diff
@@ -0,0 +1,11 @@
+if BOARD_SAMA7G54_EK
+
+config SYS_CLOCK_HW_CYCLES_PER_SEC
+	default 10000000
```
needs to come from dts property

TonyHan11 replied:
The config here will be removed and add `CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC=10000000` to `boards/microchip/sam/sama7g54_ek/sama7g54_ek_defconfig`. Thank you very much.

nordicjm replied:
no, you need to get it from a dts property, if you just move this to a defconfig you're going to get the same nack and comment

TonyHan11 replied:
> needs to come from dts property

Hi @nordicjm , 
I haven't found a way to get this config from dts. As it is described in [Kernel/Kconfig](https://github.com/zephyrproject-rtos/zephyr/blob/main/kernel/Kconfig#L828), this option is set by the SOC's or board's Kconfig file.
Might it is confused with another option.
In the support for other boards `SYS_CLOCK_HW_CYCLES_PER_SEC` come from the defconfig too (see `grep -Hnr SYS_CLOCK_HW_CYCLES_PER_SEC boards/`).
Thank you for your advice.

nandojve replied:
https://github.com/zephyrproject-rtos/zephyr/blob/4d848ebdb1148794189661b7e5741c0f96ee22b8/soc/atmel/sam/Kconfig.defconfig#L17-L18

TonyHan11 replied:
@nandojve Got it, thank you.

TonyHan11 replied:
Updated, thanks.

nordicjm commented on boards/microchip/sam/sama7g54_ek/doc/img/sama7g54_ek.jpg:
convert to webp, then put through https://tinypng.com/

nordicjm commented on boards/microchip/sam/sama7g54_ek/doc/index.rst:
```diff
@@ -0,0 +1,34 @@
+The SAMA7G54-EK board itself features connectors and expansion headers for easy
+customization and quick access to leading edge embedded features such as MIKROE
+Click boards™ and Raspberry Pi expansion header plus MIPI CSI camera.
+
```
missing hw directive block

TonyHan11 replied:
OK, `Hardware` block will be added in the update. Thank you.

nordicjm commented on boards/microchip/sam/sama7g54_ek/sama7g54_ek_defconfig:
```diff
@@ -0,0 +1,12 @@
+
+CONFIG_CONSOLE=y
+CONFIG_SERIAL=y
+# CONFIG_BOOT_BANNER is not set
```
```suggestion
```

TonyHan11 replied:
The reasons of do not output the banner are list bellow:
- As `k_calloc()` is used in the clock drivers, the initialize level of clock-controller needs be `POST_KERNEL`
- The serial for console used clock-controller for initialize, so the initialize level of it needs to be `POST_KERNEL` too
- The banner is designed to be output before `POST_KERNEL`, in this case skip it to avoid the boot stuck due to the serial is not ready then

TonyHan11 replied:
OK, I'll remove all the `k_calloc()` to make it work.

nordicjm commented on boards/microchip/sam/sama7g54_ek/sama7g54_ek_defconfig:
```diff
@@ -0,0 +1,12 @@
+CONFIG_UART_CONSOLE=y
+CONFIG_USART_SAM=y
+CONFIG_CLOCK_CONTROL_FIXED_RATE_CLOCK=y
+CONFIG_HEAP_MEM_POOL_SIZE=32768
```
```suggestion
```

TonyHan11 replied:
In code of clock drivers (`soc/microchip/sam/common`), [`k_calloc()`](https://github.com/zephyrproject-rtos/zephyr/pull/89145/files#diff-422ebc67f2216d272e90c21acb8eb4b4a72bee04438b233329c046e8a93ae0cfR162) are used for acquiring memories for struct of clocks. The default value of `CONFIG_HEAP_MEM_POOL_SIZE` is `0` and it would cause `k_calloc()` fail.

pdgendt replied:
This is the part that feels wrong, SoC code shouldn't depend on heap.

nordicjm replied:
a clock driver should frankly not need malloc at all, I would say something is very wrong with your driver design

nordicjm replied:
Ping @nordic-krch for comment

nordicjm commented on drivers/serial/usart_sam.c:
```diff
@@ -6,6 +6,10 @@
+#ifdef CONFIG_DT_HAS_MICROCHIP_AT91SAM9260_USART_ENABLED
+#undef DT_DRV_COMPAT
+#define DT_DRV_COMPAT microchip_at91sam9260_usart
+#endif
```
rename the driver instead?

nordicjm replied:
`atmel_sam_usart` just seems to be `microchip_at91sam9260_usart`?

TonyHan11 replied:
Will add a new driver for `usart` to avoid `#undef` and `#define`. Thank you for your comments.

nordicjm replied:
I was more meaning why not just use the existing name? Why does need a new binding and driver name when the driver (and, I assume silicon) is the same?

TonyHan11 replied:
The driver is the same (for USART standalone vs Flexcom (USART/TWI/SPI included) peripheral), actually I used it to handle the initialize level issue. We can use the driver after solving the `k_calloc()` issue, thanks.

nandojve replied:
> `atmel_sam_usart` just seems to be `microchip_at91sam9260_usart`?

Ahh ok.

What I understood is that maybe we need to support multiple dts bindings here. The MCHP will start to send a lot of new bindings that can be shared between Microchip/Atmel because the driver is the same. I assume this will be something common over the time.

For instance, in the past I fixed the the Ethernet and shared between to complete different series like this:
https://github.com/zephyrproject-rtos/zephyr/blob/95db3a66fd85a4faae6d09bdc9843ecba7efb872/drivers/ethernet/eth_sam_gmac.c#L22-L26

So, the outcome for me is define the acceptable pattern to allow multiple bindings inside the driver.

TonyHan11 commented:
@nordicjm @pdgendt @nandojve 

Updates the commits with:
-	Replace the image with a small webp formatted file in boards/microchip/sam/sama7g54_ek/doc/img
-	Add `Hardware` block in boards/microchip/sam/sama7g54_ek/doc/index.rst
-	Re-use drivers/serial/usart_sam.c with the same compatible keyword for sama7g5 USART
-	Remove `k_calloc()` from the code for clocks. As the result remove `CONFIG_HEAP_MEM_POOL_SIZE=32768` and `# CONFIG_BOOT_BANNER is not set` from defconfig file, adjust initialize level of drivers from `POST_KERNEL` to `PRE_KERNEL_1`
-	Rebased on the latest main branch
-	`SYS_CLOCK_HW_CYCLES_PER_SEC` kept unchanged (configured in boards/microchip/sam/sama7g54_ek/Kconfig.defconfig)

Thanks a lot for your suggestions and advice.

nordicjm commented on boards/microchip/sam/sama7g54_ek/Kconfig.defconfig:
```diff
@@ -0,0 +1,11 @@
+
+if BOARD_SAMA7G54_EK
+
+config SYS_CLOCK_HW_CYCLES_PER_SEC
+	default $(dt_node_int_prop_int,/soc/timer@e1800000,clock-frequency)
```
this should be done at soc level

TonyHan11 replied:
Updated with moving this config to `soc/microchip/sam/sama7g5/Kconfig.defconfig`, thank you.

TonyHan11 commented:
@kartben,
updated according to the comments, thanks.

nandojve commented on soc/microchip/sam/common/pmc.h:
```diff
@@ -0,0 +1,486 @@
+	unsigned int npck;
+	struct device **pchws;
+
+	struct device *hwtable[241];
```
241 comes from where ? I expect this to be defined somehow inside the SoC headers in the HAL or in another place.

nandojve commented on soc/microchip/sam/common/sama7g5.c:
```diff
@@ -0,0 +1,1226 @@
+	struct clk_range r;
+	uint8_t chgp;
+	uint8_t id;
+} sama7g5_periphck[] = {
+	{ .n = "pioA_clk",	.p = PCK_PARENT_HW_MCK0, .id = 11, },
+	{ .n = "securam_clk",	.p = PCK_PARENT_HW_MCK0, .id = 18, },
```
I think it is OK to start this way but this do not scale.
Can this structures be auto-generated from SVD or atfd files ?

TonyHan11 commented on boards/microchip/sam/sama7g54_ek/sama7g54_ek.yaml:
```diff
@@ -0,0 +1,14 @@
+arch: arm
+toolchain:
+  - zephyr
+ram: 128
```
The 128k is for the internal SRAM for the SOC and 512MB is the size for DDRAM on board.
There's no flash (NVM) in this chip. Should it be added with `flash: 0`? thank you.

TonyHan11 commented on drivers/pinctrl/Kconfig.sam:
```diff
@@ -7,3 +7,11 @@ config PINCTRL_SAM
 	help
 	  Atmel pin controller driver is used on SAM and SAM4L SoC series
+
+config PINCTRL_SAM_PIO4
```
The PIO peripheral is not compatible with the one in mainline. See [registers for samv71 PIO](https://github.com/zephyrproject-rtos/hal_atmel/blob/master/asf/sam/include/samv71/component/pio.h#L7273) and [registers for sama7g5 PIO](https://github.com/zephyrproject-rtos/hal_microchip/pull/27/files?diff=unified&w=0#diff-fbef2ca189ecb76caa625098e4f17d7fab0909e6b750daae18129a1d9a7187faR5821).

TonyHan11 commented on dts/bindings/clock/microchip,sama7g5-pmc.yaml:
```diff
@@ -0,0 +1,22 @@
+description: |
+  Microchip Power Management Controller (PMC)
+
+compatible: "microchip,sama7g5-pmc"
```
As the driver for clocks are mainly located at `soc/microchip/sam/common` this part could be a generic one. I'll update it for sharable with others, thank you.

TonyHan11 commented on dts/bindings/clock/microchip,sama7g5-sckc.yaml:
```diff
@@ -0,0 +1,23 @@
+description: |
+  Microchip SAM Slow Clock Controller (SCKC)
+
+compatible: "microchip,sama7g5-sckc"
```
This SCKC driver is dedicate for sama7g5 chip.

TonyHan11 commented:
@nandojve Thank you for your comments and suggestions. 

Updated according to the comments with the changes bellow:
- updated the frequencies with `DT_FREQ_M` / `KHZ` / `MHZ`
- align the `\` on the right side
- change compatible restrictions for `microchip_sam_pit64b`, `microchip_sam_pmc` and `microchip,sam-flexcom`
- the 64bit timer temporarily works in 32bit mode, to be improved later
- update magic number in `clocks_gck[46]` to `ID_PERIPH_MAX`
- simplify code with variables in `clk-generated.c`
- remove const values defined in `soc/microchip/sam/common/pmc.h`, use definitions for PMC registers in hal instead
- remove `struct device *hwtable[241];`,  add `static struct device *pmc_table[PMC_MCK1 + 1 + CLK_CNT_SYSTEM + CLK_CNT_PERIPH + CLK_CNT_GCK + 8];` for the same purpose

nandojve commented:
Hi @TonyHan11 ,

Changes seems Ok, thank you for your effort to bring this in.
Let's just make CI happy : )

nandojve commented:
BTW, this is a side question;

What is necessary after this to make Zephyr work on a SAMA5D3/D4 ?
Is there any plans to make those work ?

TonyHan11 commented:
> BTW, this is a side question;
> 
> What is necessary after this to make Zephyr work on a SAMA5D3/D4 ? Is there any plans to make those work ?

@nandojve The draft plan is to add more peripheral support for sama7g54 first, then adding support for SAMA7D65, SAMA5D2 and so on. Thanks a lot.

TonyHan11 commented:
@nandojve One question for the PR, there are some failing checks (`Run test with twister`) for the commits, need I fix them all before the PR been merged? Thanks a lot.

nandojve commented:
Hi @TonyHan11 ,

> @nandojve One question for the PR, there are some failing checks (`Run test with twister`) for the commits, need I fix them all before the PR been merged? Thanks a lot.

The issues came from your PR and they need to be solved, yes.

TonyHan11 committed:
west.yml: Update west.yml for sama7g54 hal

Update west.yml for SAMA7G54 HAL

Signed-off-by: Tony Han <tony.han@microchip.com>

TonyHan11 commented:
@nandojve
updated with the revision for hal_microchip and fix the error in CI checks.
Result: `1 neutral, 2 skipped, 26 successful checks`.

TonyHan11 commented:
@kartben , @nordicjm , 
Would you like to review the pull request again as all the known issues have been fixed? Thanks a lot in advance.

kartben commented:
Let's wait a bit for @nordicjm to confirm he's +1 on the additional changes but let's proceed to merge otherwise as he had previously approved
